### PR TITLE
Feature/seq num

### PR DESCRIPTION
**Pull request by xuyao0127**

### CODE REVIEW

File: mpi-proxy-split/seq_num.cpp [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+
+#include "jassert.h"
+#include "seq_num.h"
+#include "mpi_nextfunc.h"
+#include "virtual-ids.h"
+#include "record-replay.h"
+
+using namespace dmtcp_mpi;
+
+#define DEBUG_SEQ_NUM
```
Comment (JainTwinkle) [inline mpi-proxy-split/seq_num.cpp:15]: Quick comment: I guess this should be commented out by default.

File: mpi-proxy-split/test/Makefile [hunk 2]
```diff
@@ -32,8 +32,9 @@ FILES=mpi_hello_world \
       Comm_get_attr_test Initialized_test Cart_sub_test\
       Barrier_test Bcast_test Comm_split_test Reduce_test \
       Waitall_test Waitany_test Testany_test \
-      two-phase-commit-1 two-phase-commit-2 Cartdim_get_test \
-      Ibcast_test Ibarrier_test Isend_test Cart_map_test \
+      two-phase-commit-1 two-phase-commit-2 two-phase-commit-3 \
+      Ibcast_test Ibarrier_test Isend_test
+      Cart_map_testi Cartdim_get_test\
```
Comment (JainTwinkle) [inline mpi-proxy-split/test/Makefile:37]: space before `\`

File: mpi-proxy-split/seq_num.cpp (hunk 1 above)
Comment (xuyao0127) [reply mpi-proxy-split/seq_num.cpp:15]: Yes. I will comment it out.

**Comment by gc00** [comment]
@xuyao0127 , Could you reply to @dahongli about when the target is reached?  After that, @dahongli can approve.
Thanks.

File: mpi-proxy-split/seq_num.cpp [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+
+int check_seq_nums() {
+  unsigned int comm_id;
+  unsigned int seq;
+  int target_reached = 1;
+  for (comm_seq_pair_t pair : seq_num) {
+    comm_id = pair.first;
+    seq = pair.second;
+    if (target_start_triv_barrier[comm_id] > seq_num[comm_id]) {
+      target_reached = 0;
```
[the remark this answers is not in the document]
Comment (xuyao0127) [reply mpi-proxy-split/seq_num.cpp:84]: I agree it should break if the condition is true. But this condition is "there exists a communicator which haven't reached the target.

File: mpi-proxy-split/mpi_plugin.cpp [hunk 4]
```diff
@@ -323,12 +324,20 @@ mpi_plugin_event_hook(DmtcpEvent_t event, DmtcpEventData_t *data)
         while (1) {
           // FIXME: see informCoordinator...() for the 2pc_data that we send
           //       to the coordinator.  Now, return it and use it below.
-          rank_state_t data_to_coord = drainMpiCollectives(coord_response);
+          rank_state_t data_to_coord = preSuspendBarrier(coord_response);
```
[the remark this answers is not in the document]
Comment (xuyao0127) [reply mpi-proxy-split/mpi_plugin.cpp:321]: Fixed.

File: mpi-proxy-split/seq_num.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (xuyao0127) [reply mpi-proxy-split/seq_num.cpp:84]: Fixed

**Review by dahongli — APPROVED** [review]
Thanks Yao! It looks good. 

Please feel free to merge into main branch. Twinkle is going to tag the main branch and save that into a new branch in case some people wants to continue using that code base.

**Comment by JainTwinkle** [comment]
@xuyao0127 @dahongli @gc00 
I have created a new tag (`MANA-2PC`) and a branch (`dev/jaintwinkle/main-2pc-backup`). I guess you can merge your PR into the main.

**Pull request merged**